### DISCUSSION TITLE
How do I use SwiftUSD with Xcode based projects?

### DISCUSSION THREAD

**Discussion by x001017** [wabiverse/swift-usd#20] [Q&A] [answered]
I created an Xcode project and installed SwiftUSD with Swift Package Manager, which is linked below.

However, after installing,

import PixarUSD

and tried to build the project with the default code when newly created, but before the code is built, many errors occur as follows.

Multiple commands produce '/Users/name/Library/Developer/Xcode/DerivedData/SwiftUSDPackageCamera-gpbofwclcozunoespbordaiizrfp/Build/Products/Debug-iphoneos/Frameworks/Alembic_301B1C71A317B7B4_PackageProduct.framework'

Target 'Examples' (project 'SwiftUSD') has copy command from '/Users/name/Library/Developer/Xcode/DerivedData/SwiftUSDPackageCamera-gpbofwclcozunoespbordaiizrfp/Build/Products/Debug-iphoneos/PackageFrameworks/Alembic_301B1C71A317B7B4_PackageProduct.framework' to '/Users/name/Library/Developer/Xcode/DerivedData/SwiftUSDPackageCamera-gpbofwclcozunoespbordaiizrfp/Build/Products/Debug-iphoneos/Frameworks/Alembic_301B1C71A317B7B4_PackageProduct.framework'

Multiple commands produce '/Users/name/Library/Developer/Xcode/DerivedData/SwiftUSDPackageCamera-gpbofwclcozunoespbordaiizrfp/Build/Products/Debug-iphoneos/Frameworks/Alembic_301B1C71A317B7B4_PackageProduct.framework/Alembic_301B1C71A317B7B4_PackageProduct'.....

In the Swift Package Manager,
“If there are files with the same name in multiple directories, each with the same name, an error will occur during build.”
which is what the search says about Swift Package Manager errors. Does the fact that such errors occur mean that I need to modify the file names after installing with Swift Package Manager, for example by changing them?

**Comment by furby-tm** [comment]
@x001017

> “If there are files with the same name in multiple directories, each with the same name, an error will occur during build.”

This is generally correct, for files (`SomeFile.swift`) with the same name in multiple directories **_within the same project_**. However, I don't believe this is the specific error you are running into here, at least from the error output you provided above.

The error output you provided above appears to indicate that you have added some other framework not related to the SwiftUSD package to your project, perhaps some sort of private or third-party framework you added for Alembic?

The SwiftUSD package already brings in an [**Alembic** dependency from the **MetaverseKit** package](https://github.com/wabiverse/MetaverseKit/tree/main/Sources/Alembic) which is automatically compiled from source and linked against the **`PixarUSD`** product when you build a project that depends on **`PixarUSD`**, so you should not need to add your own external `Alembic` framework to your project as the `Alembic` dependency has already been provided for you.

Once you've removed your external `Alembic` framework from your project, it's generally advised to:

1. Quit Xcode
2. Delete Xcode's `DerivedData` cache
3. Open Xcode and build your project.

To ensure the build graph is cleanly refreshed to no longer expect the external Alembic framework that you had initially added to your project.

**Reply by x001017** [reply]
Thanks for the reply.
This time I only installed SwiftUSD with Swift Package Manager and no other Alembic related frameworks.
I have also tried clearing the DerivedData cache in Xcode, but it didn't make a big difference.
Maybe,
Is it possible that a mistake exists in the way SwiftUSD is introduced in the Swift Package Manager? I am also thinking that there may be a mistake in the way SwiftUSD is installed in the Swift Package Manager.

The way I did it this time was to create a new project, leave the defaults untouched, and in SPM, use the
“https://github.com/wabiverse/SwiftUSD.git”
and set Dependency Rule to “Brunch”, “main”, “Add to Project”, and “Add Package” to the project you want to install.
Package installation starts, and in the selection, Ar, Arch, CameraUtil, PixerUSD, etc. are all selected, and Add.
If I run the build in that state, the same problem occurs. Does this mean that this installation method is already wrong? Sorry for the rudimentary question.

**Reply by furby-tm** [reply]
It's hard to tell what's going on without seeing an example of your `Package.swift` file, but what I can say is that I tested the following on macOS, visionOS, and iOS and did not experience any issues:

```swift
// swift-tools-version: 5.10

import PackageDescription

let package = Package(
  name: "USDTesting",
  platforms: [
    .macOS(.v14),
    .visionOS(.v1),
    .iOS(.v17),
    .tvOS(.v17),
    .watchOS(.v10)
  ],
  dependencies: [
    .package(url: "https://github.com/wabiverse/SwiftUSD.git", from: "24.8.7")
  ],
  targets: [
    .executableTarget(
      name: "USDTesting",
      dependencies: [
        .product(name: "PixarUSD", package: "SwiftUSD")
      ],
      swiftSettings: [
        .interoperabilityMode(.Cxx)
      ]
    )
  ],
  cxxLanguageStandard: .cxx17
)
```

Which version of macOS/(and iOS, if you are building for iOS) and Xcode are you using as well?

**Reply by furby-tm** [reply]
Are you also using a `.xcodeproj` (Xcode project) by chance? If so, it sounds like you have some left over Xcode configuration that's causing you some problems.

> [!TIP]
> I will note here, that I personally avoid `.xcodeproj` across all my projects, projects which are wholly dependent on Xcode tend to add quite a lot of developer friction, and my preference is using only a single `Package.swift` file at each project's root directory.
>
> This allows me to _optionally_ use Xcode, and additionally distribute my apps and libraries across all Apple devices (**macOS**, **iOS**, **visionOS**, and **tvOS**), **Linux**, and **Microsoft Windows**, while also leveraging code editors like Visual Studio Code. I do this with [**SwiftBundler**](https://swiftbundler.dev), which currently has support for **macOS**, **iOS**, **visionOS**, and **tvOS**, for all other non-Apple platforms I just interface with `swift build` from the command line.

In the following screenshot here, do you by chance have anything in your **Frameworks Libraries & Embedded Content** section? Try removing everything listed in there, if you see anything listed. Though, if you are still experiencing issues, I would advise setting up a brand new project for this.

<img width="70%" src="https://github.com/user-attachments/assets/0ed24e13-1d0f-447e-8896-41ebc8c894f3" />

**Reply by x001017** [reply]
The OS used is Sonoma 14.7.1.
Xcode is using 15.4.

The work is done by creating a new project on Xcode.
Would the method using Xcode still not work?

After creating a new project on Xcode, I immediately add SwiftUSD with SPM, but then manually create Packaging.swift and write it there, is that correct?
Create a new project and do what you did above.

https://github.com/x001017/USDTesting

I have zipped the file to
I hope you can take a look at it.

**Reply by x001017** [reply]
What do you think?
Did the above zip code work?
If we remove the framework stuff, SwiftUSD will not work at all, and we are currently stuck.

**Reply by furby-tm** [reply]
@x001017 sorry about the long wait, I've been out of town on vacation, I will review your project and Xcode based projects (as opposed to swiftpm ones) when I return, I appreciate your patience.

**Comment by furby-tm** [comment, marked as answer]
@x001017 going through Xcode now (through **Xcodeproj**, as opposed to the **SwiftPM** route), wow! **Xcode** really overcomplicates the simplicity of SwiftPM's package manifest system.

Hidden through various settings, I believe you should get it to work.

1. Create a new app in Xcode.

2. Add the SwiftUSD package to your Xcode project like so:
   <img width="500" alt="Screenshot 2024-11-24 at 1 34 33 PM" src="https://github.com/user-attachments/assets/023c3416-b81d-48cf-b1c0-6136bfb835f9">

3. Additionally you must add the **PixarUSD** product to your app's target.
   <img width="674" alt="Screenshot_2024-11-24_at_1 17 05_PM" src="https://github.com/user-attachments/assets/e410b6e0-11b0-4c53-945b-4dd38e748df3">

4. A very important step, and one that is way too hidden, you need to enable C++ interoperability mode here:
   <img width="757" alt="Screenshot_2024-11-24_at_1 29 09_PM" src="https://github.com/user-attachments/assets/6d5a3e3b-be1c-4191-aeca-4c7cad93cbbf">

5. Equally as important, you need to ensure you set the Swift version to 6, here:
   <img width="1191" alt="Screenshot 2024-11-24 at 1 38 41 PM" src="https://github.com/user-attachments/assets/8cef9854-1b72-4a8b-8e98-2a3e6fac0d85">

6. You need to change yet another very hidden setting buried in your Xcode project, which is to set the C++ Language Standard to c++17:
   <img width="1195" alt="Screenshot 2024-11-24 at 1 50 09 PM" src="https://github.com/user-attachments/assets/9437b375-a740-4fbe-8e32-78ba7b6dda27">

I am sorry that Xcode's poor development experience had blocked your team's software development. Hopefully, once you follow the above steps, things can begin moving along smooth for you.

**Reply by x001017** [reply]
Thanks for your reply. Very helpful, I see that it was not available in Xcode 15.4, I will try to work on Xcode 16 to be able to configure the project.
It really helped me a lot. In the meantime, I would like to make a personal challenge.

**Comment by furby-tm** [comment]
<img width="1378" alt="Screenshot 2024-11-24 at 1 56 10 PM" src="https://github.com/user-attachments/assets/6983c3c7-2fa3-4582-9b78-57c9d9a6b967">

And I built the project to confirm that these are the steps that worked for me.
